### DISCUSSION TITLE
Support to activate WiFi, change flash to 16Mb

### DISCUSSION THREAD

**Discussion by bixb922** [bixb922/crank-organ#3] [General]
Hello,
I'm answering a post from another forum here:
>I'm using a Raspberry Pi 5 with Linux as the PC.

Very interesting! I have only used a Windows PC and now a Mac, and I think using a Raspberry Pi 5 as a PC is amazing! Great you worked that out!!

>No wifi connection was successful, on either SSID, up to the 15-minute timeout

The log shows that the microcontroller is connecting alternately to a router or hotspot called ```wifi_SSID_2```and ```wifi_SSID_1```. Status 201 means "No access point found".  These are the default names. 

>My phone didn't detect an esp32 network

I checked again today. I installed the software on a "fresh" ESP32-S3. Then I selected the option to see all available WiFi networks on my cell phone and on my PC. On the list of available WiFi networks I can see ```esp32s3```, can you see that too? This WiFi network is published only for about 2 minutes after reboot and goes away if nobody connects to save energy. You can connect with a cell phone or a PC, and will be asked the standard password which currently is ```password```.  This way to connect is provided mainly for initial configuration.

>I'm wondering if the drivers all installed properly? 

The drivers are installed together with MicroPython. If they were not, you would get message that the ```network``` module is not available. To check the Wifi, please use ```mpremote``` to get the REPL prompt, i.e. the ```>>>``` prompt. You are now connected to MicroPython and can enter Python statements. Enter the following:
```py
import network
x=network.WLAN(network.STA_IF)
x.active(True)
x.scan()
```
You should get a list of available WiFi networks where the microcontroller can connect.

>I haven't tried the "mp-image-tool". The instructions left me so confused I had no idea what to do. So my 16MB board may be accessing just 8MB. (esptool reports the board has 16MB)

Yes, 8Mb will work also, but with less capacity for tunes. With the latest version of the ```mp-image-tool-esp32``` things are easier. This is what I did this today to convert the MicroPython image to 16 Mb capacity:  At the command prompt (in your case, on the Raspberry Pi 5), install the mp-image-tool-esp32 utility:
```
pip install mp-image-tool-esp32
```
Now you should put the ESP32-S3 in firmware update mode, i.e:
* Hold down the small BOOT button on the board
* Press the small RST button on the board for a short time
* Now release the BOOT button

Now you can convert a MicroPython .bin image and load it to the board in one go, what worked on my Mac is:
```
mp-image-tool-esp32 ESP32_GENERIC_S3-SPIRAM_OCT-20241129-v1.24.1.bin -f 16M --resize vfs=0 --flash /dev/cu.usbmodem14201 
```
The .bin file is the latest release of MicroPython, downloaded from their site. ```-f 16M``` means "set flash size to 16M", and "--resize vfs=0" resizes the file system. ```--flash```does the same job as ```esptool.py``` and sends the image without any further step to the ESP32-S3.  ```/dev/cu.usbmodem14201 ```is the USB serial port on my Mac where the ESP32-S3 is connected, you have to substitute with the port on your Raspberry Pi.

I wasn't aware of this latest version of the ```mp-image-tool-esp32```. It has improved a lot, there are lots of options now. ```esptool.py``` isn't needed anymore, and it even does some ```mpremote``` commands.

Connecting to the ```esp32s3``` allows to configure the WiFi network properly, inserting the names of your home WLAN. You also could do this to configure the WiFi on the Microcontroller:
* Copy the configuration file from the microcontroller to the Raspberry Pi with ```mpremote cp :/data/config.json ./``` 
* Edit that file with ```vi``` or some text editor of your choosing and set the values of ```access_point1```, ```password1```, ```access_point2``` and ```password2``` to your home WLAN and cell phone hotspot respectively, 
* Copy the configuration file back with ```mpremote cp config.json :/data```
Best regards, hope this helps one step further. Please post how it goes.

**Comment by viragored** [comment]
Fantastically helpful information - thanks!! Quick recap of my response on the other forum:

- wifi working fine, I didn't realise it would come and go while I was watching the setup log messages scrolling
- web pages appearing as they should on 192.168.144.1
- device connected to house wifi so is reachable from desktop PC

Following the instructions [here](https://micropython.org/download/ESP32_GENERIC_S3/) I tried these:

- Erased the flash memory on the board, and reinstalled Micropython latest version. That all worked fine.
- Used mp-image-tool-esp32 (without erasing flash) as you showed above - failure, with messages about corrupt memory.
- Erased the flash again, used the mp-image... tool again - failure, same messages about corrupt memory.
- Reverted to the working setup with 8MB, the first step again

A difference I noticed between the micropython instruction using esptool and the mp-image... instruction is that the former specifies a base address of zero. I've now run out of time to explore more today, so I'll get back to this some time later.

And this is the most amazing piece of software - you've created something wonderful!

**Comment by bixb922** [comment]
>device connected to house wifi so is reachable from desktop PC

That is great news! Excellent, that makes configuring easy.

>Used mp-image-tool-esp32 (without erasing flash) as you showed above - failure, with messages about corrupt memory.

Is this the message?
```
The filesystem appears to be corrupted. If you had important data there, you
may want to make a flash snapshot to try to recover it. Otherwise, perform
factory reprogramming of MicroPython firmware (completely erase flash, followed
by firmware programming).
```
This would indicate that MicroPython is running ok, but the partition with the file system is corrupt, as it may well be after resizing the flash memory with the mp-image... tool.
If this is the message, please press control-C at this point. You should get the REPL prompt, i.e. ```>>>```. At the prompt, enter:
```
import os
os.VfsLfs2.mkfs(bdev)
```
Now reboot the microcontroller, for example pressing control-D or pressing the RST button. The VfsLfs2 command should have reformatted the file system. To check MicroPython now sees the 16MB of memory, enter the following at the REPL prompt:
```
import os
os.statvfs("")
```
The output should be ``` (4096, 4096, 3584, 3582, 3582, 0, 0, 0, 0, 255)```. That means that you have 4096*3582=14,671,872 bytes of memory available (16Mb minus 2Mb used by MicroPython). For 8Mb flash, you would see ```(4096, 4096, 1536, 1534, 1534, 0, 0, 0, 0, 255)``` meaning 6,283,264 bytes available.

Did that work?

Thank you very much for the nice comments. Let's get your organ working now, to justify those nice comments :-) good luck!

**Reply by viragored** [reply]
Once more, huge thanks for the effort you've put into replying so comprehensively. Long story short - the board now has 16MB memory, your software is installed and we're ready to rock and roll. 
I kept detailed notes as I went through your steps. Since everything worked as expected you probably won't want to read the notes.
[esp32-s3 Memory expansion.txt](https://github.com/user-attachments/files/18123632/esp32-s3.Memory.expansion.txt)

I'm still woodworking, very slowly, but hope to soon be trying out the electronic components in breadboard-style. Power will be from a bench power supply, feeding an [Orgautomatech decoder board](https://orgautomatech.fr/produit/midi2org-32/), MIDI music coming from the esp32-s3. I expect I'll need the Pico translator between the two devices, but haven't looked into that aspect yet.
I will need to shop some MIDI plugs to fit the decoder, which might have to wait until after Christmas.

Many thanks once more :-)

**Comment by bixb922** [comment]
Great! I read the attached text file, everything seems ok!

>mpremote finds no device [expected behaviour, board not in bootloader mode]

That's probably because Thonny is already connected to the board. To use mpremote, you need to stop Thonny. While the board is in boot mode, it usually has another port, so mpremote can now see the board.

>a static IP address and all seems well.

The IP address is probably dynamic, i.e. assigned by the router. Normally, home routers always assign the same IP address to the same board, based on the MAC address of the board, so it behaves like a static address. I found that handy, because on older Android phones the name (i.e. "organ.local") does not work in the browser, but the IP address always work (i.e. http://192.168.100.57/ or whatever). You can see the assigned IP address in the startup log or with "System" button.

**Reply by viragored** [reply]
As I was taking baby steps with this process, I deliberately left Thonny active so I could confirm progress and (in a roundabout way) know if the board was in boot mode or not. 
My router's DHCP server doesn't assign the same dynamic address every time, so I've put my chosen static address in the router's reserved table. That way I can get straight to the home page with a bookmark. In the Windows 10 and Linux worlds, the ".local" address seems rather hit and miss.
Exciting times are coming, I'm just about to take the clamps off the freshly glued bellows!

**Comment by viragored** [comment]
Hi again. I'm ready to try a bench-test of ESP32-S3 midi out to Orgautomatech decoder board. Can I just check with you to make sure I'm doing the right things?

- "64 Notes MIDI over serial" selected on the ESP
- default GPIO 10 for Tx
- ESP's GP10 connected to Board MIDI Rx
- ESP's Gnd to Board Gnd
- Play a MIDI file via the web interface

Is there anything else to connect between ESP and Board?
ESP will be powered by USB (wall socket), Board by 12V bench power supply

I have a test setup with the board driving valves, MIDI output from Raspberry Pi 5 via a USB to MIDI serial converter. And I've created a test file in Cakewalk (Windows PC, not Pi) to test the valves individually and together, so I'm feeling quite pleased that it's time to bring the ESP into play.

![Test 10 solenoids copper tape (2)](https://github.com/user-attachments/assets/7021ee28-606f-4456-b8fa-63f4f29891d8)
Also I've bought from AliExpress a rotary encoder https://www.aliexpress.com/item/1005007061739358.html 
It's 100P (they had no 200P models when I ordered) so I hope it will be OK.

**Comment by bixb922** [comment]
Hello,
>Is there anything else to connect between ESP and Board?

MIDI is activated by current, not by voltage. The receiving circuit is a optocoupler.  The current to the optocoupler has to be limited. Connecting the ESP32 GPIO directly could draw too much current.

These are the official specs for MIDI: https://midi.org/5-pin-din-electrical-specs
There is a recommended diagram on page 2. The detailed calculations that lead to the resistor values are on page 5.

The connection should be (according to the standard):

```mermaid
graph LR;
GPIO10-->Ra[33Ohm]-->DIN4
GND --> Rc[10Ohm] --> DIN5
```
According to the calculation in the standard, the current that the ESP32 would have to source is about 5mA, which well within the capabilities of the ESP32.  The polarity is important.

I believe this should also work:
```mermaid
graph LR;
GPIO10-->R[47Ohm]-->DIN4
GND --> DIN5
```

Unfortunately I have no Organautomatech board, and can't test this.


>Also I've bought from AliExpress a rotary encoder

Yes, that's the same model I am using. 100p should work perfectly well. You have to set that value in the "General configuration" page. Both output A and B of the sensor should be connected. On the "Pin and MIDI configuration" page you need to specify the two pins.

>I have a test setup with the board driving valves

Looks nice!!

**Comment by viragored** [comment]
Thanks once more for your generous and awe-inspiring help. After a couple of weeks away from this I've been ploughing through the info on MIDI standard. The bad news is I'm fairly confused! 
The good news is the Orgautomatech board is fully MIDI compliant, so things like opto-couplers are in place. The board has a socket for a standard MIDI plug, out from the esp and in to the board. And I've discovered that I need a logic level converter to raise GPIO 3.3V to MIDI 5V. When some of those have arrived I'll do a breadboard setup and see if I can make the esp rattle some valves.
A couple of quick questions:

- Do I have to somehow connect the esp to the shield that is normally on pin 2, or can I ignore that?

- And just for my own peace of mind - is the GPIO output +ve? (I think it must be, but I'd welcome assurance that I've got it right)

I've drafted a block diagram of the electrical and electronic wiring I think I'll need. (The resistors in the MIDI connection are not shown yet.) If you happen to spot any basic blunders on that, I'll be pleased to have them pointed out!

While experimenting with the valves on the test board I've measured the voltage across the valves at ~12V with just one valve operating, but ~10.2V with all ten operating. So I think I'll need to bring the input voltage up to about 13.5V to assure their effective operation under wind pressure.
![Electronics schematic](https://github.com/user-attachments/assets/0c761383-6d51-4a42-a779-d7b2a5ecd3ea)

**Reply by bixb922** [reply]
Nice progress! Glad to help.

>And I've discovered that I need a logic level converter to raise GPIO 3.3V to MIDI 5V.

Take much care here. MIDI is a current loop, it needs 5mA, and the voltage to get this is around 1.2V.  You can use a 5V level shifter, but you need a 220 Ohm resistor connected between level converter and DIN plug pin 5 and another 220 Ohm resistor connected between level converter and DIN plug pin 4.

The resistors need for a 3.3V driver such as the ESP32 output are much lower, they are 33Ohm and 10Ohm, as before, in series with pin 4 and pin 5 of the DIN connector.

>I've drafted a block diagram of the electrical and electronic wiring I think I'll need. (The resistors in the MIDI connection are not shown yet.) 

Please post that with the resistors in the MIDI connection, I'll be happy to verify. What's posted looks well to me. Common ground is ok.

I have ordered DIN plugs and some components, and I just haven't found the time to test the MIDI out circuit. I'll keep you updated as soon as I can test that. I'll try first to do it without a level converter.

>Do I have to somehow connect the esp to the shield that is normally on pin 2, or can I ignore that?
If the cable is short, say a few centimeters, no shield to pin 2 is necessary.


>And just for my own peace of mind - is the GPIO output +ve?
+ve meaning that the GPIO output is positive with respect to ground? Yes, that is correct.

>~12V with just one valve operating, but ~10.2V with all ten operating. So I think I'll need to bring the input voltage up to about 13.5V to assure their effective operation under wind pressure.

With 13.5V for sure you'll have no problems. Have you been able to measure the pressure of the bellows? What is that pressure? And what is the diameter of the largest toe hole?

When I built the first wind chest, I used ten rechargeable NiMh AA batteries without converter. The nominal voltage is 12V and it decays when the batteries loose charge and voltage can drop to 10.8V. The air pressure is around 8cm H2O.he larges toe hole is 11.2 mm in diameter. This worked well, no problems with the voltage..

However, when using this with all pipes, I ran into problems. The largest toe hole is 15.7mm (bass pipes), and with a lower voltage, these bass valves are very slow to open with 12V. A voltage boost converter to 13.5V will made the valve response very stable. Peterson valves will have no problem whatsoever with 13.5V. Just make sure the DC-DC voltage boost converter can handle 2 Amperes.

The software has a parameter to limit the number of open valves (polyphony) so this is another safeguard to ensure a maximum current is not exceeded.

**Reply by viragored** [reply]
Thanks once more for being so helpful. And yet again I thought I had understood something ("MIDI standard requires 5V") but didn't (a current loop is required). I don't understand what "current loop" means or implies, so I'll spend some time with Google and see if I can educate myself.

Lately I've not had time to get back to organ stuff and when the bits I've ordered (DIN plugs and logic level converters) arrive from Ali I'll do some proper experimenting. Meanwhile...

The John Smith Busker organ is designed to use 125mm (5") H2O air pressure. The large holes are all 8mm diameter with 15mm pallets.
Scott Peterson was very helpful when I wanted to order valves. From the John Smith information he recommended >12V (eg 14V) to deliver satisfactory voltage at the valves, and 60 Ohm coils although 90 Ohm _might_ work (his emphasis).

I have a 6A boost converter. When the other bits arrive I'll try a breadboard connecting GP10 and Ground from the ESP32-S3 to a DIN plug with resistors as you suggest and see what happens when plugged in to the Orgautomatech board. If nothing happens, I'll then try adding a logic level converter and resistors. 
Hopefully I know enough to not make smoke!

**Reply by viragored** [reply]
I've done a bit of Googling and reading, and have been reminded of Ohm's Law - which I last heard of at school, more than a few decades ago. If I've followed you correctly, this picture shows what I should do. 
I'd really appreciate your checking this before I connect the bits together. For breadboard testing I'll use USB power before trying 5V power to the "in" pin.
Being delighted that AliExpress had set a new delivery speed record (6.5 days) I was just about to start a breadboard setup, realised I'd misread your resistor values and thought I'd better have a double-check. That could have been a big whoopsie.
![ESP32-S3 - MIDI](https://github.com/user-attachments/assets/91d975a0-244b-4be3-8b7d-8c8c62de033e)

**Comment by bixb922** [comment]
I think that circuit would work, but the signal would have to be inverted (can be done in MicroPython).

In the meantime, I finally could spend some time setting up a MIDI over serial connection successfully. I have connected a sound synthesizer via MIDI to the ESP32-S3. It took me some time to make it work, but finally sound came out of the synthesizer and no components were burnt in the process :-) Here are the details:

https://github.com/bixb922/umidiparser/blob/master/midi_over_serial/midi_over_serial.md

I'd be very happy to get some feedback.

**Reply by viragored** [reply]
Getting music out of your esp32 and into your synth must have been a great feeling! Congratulations - and many thanks for the new documentation.

I think I understood your linked details as far as the 'Program' section, and then I got lost. I think I can see what you mean by 'inverting the circuit'. I don't think I should consider altering any MicroPython, I should assemble my hardware to match your code.
Hopefully my revised diagram interprets correctly your MIDI Circuit. But it's quite possible I've got the two connections to pins 4 and 5 the wrong way round. 
If you would be able to confirm or correct my drawing, I'll put a breadboard test together.

The only worthwhile feedback I feel confident to give you is that the pin numbering on your 'Circuit' MIDI plug, and in the narrative, doesn't seem to follow the MIDI standard and its 180° DIN plug. With Pin 1 at bottom left as you've shown it, the clockwise numbering could be 1 - 4 - 2 - 5 - 3, or if we're seeing the other side of the  plug 3 - 5 - 2 - 4 - 1 
I've spent many, many hours being confused by the hundreds of hits Google has given me for MIDI plug pin numbers and feel this may be one little thing I'm close to understanding!
![ESP32-S3 - MIDI](https://github.com/user-attachments/assets/ff82d276-c482-4b89-b66e-16f09807e614)

**Comment by bixb922** [comment]
>Getting music out of your esp32 and into your synth must have been a great feeling! 

Oh yes, that was fun. I was far too eager to see how it works, very suddenly the crank organ went from "I can make one flute sound" to "there is music coming out of the thing"!! But I should have been much more careful with the pipes. Although the crank organ sounds well, I am considering to building the pipes again to have a similar intonation for all.

>I think I understood your linked details as far as the 'Program' section, and then I got lost.

I'm sorry. In the section titled tinkering just I tried to document some things I tested.
 
>Hopefully my revised diagram interprets correctly your MIDI Circuit

Yes, I think that is correct. I had *many* gripes with the pin numbering. There are so many diagrams on internet that are plain wrong. I included some links in the documentation with correct plug numberings. The numbering on your diagram seems to be the solder side of a male plug, right? It's ok to use a male plug, that one will go directly into the MIDI IN of the board.

In any case, if you reverse pins 4 and 5 by accident, nothing bad will happen neither to the Organautomatech board nor to the ESP32-S3.

>The only worthwhile feedback I feel confident to give you is that the pin numbering

Thank you very much!! That certainly was an error and part of my confusion. I have now updated the document and I sincerely hope that I got it right!

>I've spent many, many hours being confused by the hundreds of hits Google has given me for MIDI plug pin numbers 

Me too. And I still have to look thrice and am still unsure if I got pin numbering right!!! I started testing with a wrong pin numbering diagram on internet, and it took me many hours to make the circuit work. Finally I got lucky reversing the pins 4 and 5...

Good luck!

**Comment by viragored** [comment]
Thanks, that's encouraging! As soon as I can get a couple of hours I'll get going on the breadboard.
Meanwhile, here's my favourite internet image of a MIDI-In socket with pin numbers.
![midi socket labeled](https://github.com/user-attachments/assets/fd4655b3-d106-42bc-9dbb-14953f23be85)
I find that easy to understand.

My diagram did have a label to show what the plug orientation is, but that might be hard to read on a small screen. So I've bolded the font and made it a bit larger. The rest of the diagram is all unchanged.
![ESP32-S3 - MIDI](https://github.com/user-attachments/assets/e2fe19d8-14ef-4320-9646-3f2271d7d906)
Exciting times are coming over the weekend :-)

**Comment by viragored** [comment]
So near... !!

I've wired up a cable with resistors, connected the ESP-S32 to the Orgautomatech board, and played my test music. MIDI data is travelling down the wire to the decoder board, but I get the red LED warning that data is out of range. No valves moved, but the red warning suggests data is probably being read correctly, so we're nearly there.

In the ESP32 general settings page the default MIDI channel is 0. Changing that to 1 (which the board expects) and clicking 'Save' got this error message:
![MIDI Channel error message](https://github.com/user-attachments/assets/5681f79e-2927-4ec1-9448-0519b91df507)

Any suggestions for how to fix that error?

Background info: The Orgautomatech decoder board is programmable, but MIDI channel is limited to the range 1 - 16.

**Comment by bixb922** [comment]
> clicking 'Save' got this error message:

I am sorry about that. That's due to an old version of 64_note_midi_over_serial.json, that does not match the software and had no "pin numbers". Please get the current version with ```git clone``` (is that the command you use to get the GitHub repository?) and then update software _and_ data with the following commands using the ```install_software.py``` and ```install_data.py``` just downloaded from GitHub:
```
mpremote run install/install_software.py
mpremote run install/install_data.py
```
The "General Configuration" will not be overwritten (i.e. WiFi configuration will be preserved). Use the "Pinout and MIDI definition" page to select "64 notes MIDI over serial data" again and press "Save scale", because the install_data will reset that definition.

When you load the "Pinout and MIDI definition" page you should see:
* UART = 1
* GPIO pin for tx	= 10
* MIDI channel (1-16) = 1

> , but I get the red LED warning that data is out of range. 

The Organautomatech manual says "When it receives a note on the right channel but out of range (too low or too high) the red LED flashes."

I agree that most probably the channel number does not match. The  "64 notes MIDI over serial data"  definition that was provided uses channel 1. You now should be able to define other channels on the "Pinout and MIDI configuration" page, reset the microcontroller and check.

However, the if the note is out of range, then I don't know what's wrong. The  "64 notes MIDI over serial" should match the factory note definition of the Organautomatech board. With the definition provided, the software sends all MIDI notes in the MIDI file (regardless of the original channel number in the file) of the range MIDI number 48 (C3) to MIDI number 111 (Eb8) to the Organautomatech board, changing the original channel to the channel set in the "Pinout and MIDI" definition.

A test that you can make is enter the "Tuner" page and press "Play scale" and/or "Test all pins".

"Play scale" will send all MIDI notes 48 to 111 to the board, on the defined channel. That should at least move some solenoids. It will play the scale up and down several times at increasing speed.

"Test all pins" will do exactly the same thing, but it operates at a more physical level, the result should be the same in this case.

As an additional test, this short program plays middle A (440Hz) on all channels. Save that to a file, say ```test.py``` on the PC and execute on the microcontroller with ```mpremote run test.py```

```py
from machine import UART
from time import sleep
uart = UART( 1, baudrate=31250, tx=10, rx=9)
for channel in range(16):
    print("Now playing A 440 MIDI 69 on channel", channel+1)
    uart.write( bytes( [0x90+channel, 69, 127] ))
    sleep(1)
    uart.write( bytes( [0x80+channel, 69, 0] ))
    sleep(1)
```

I hope this helps. Good luck!

**Reply by viragored** [reply]
Result! Your updated software works like a charm and 10 of the valves are tapping a little tune on their test bed.

https://github.com/user-attachments/assets/b0cf437b-03a1-46cf-bda8-7ba4cfafaea6

Here the ESP32-S3 is plugged into a USB wall socket and controlled from a browser on the Raspberry Pi5, over the house wifi. 
GP10  and 3.3V are connected through a resistor each to the DIN plug inserted in the decoder board (as in my diagram dated 27.3.25) so there is no need for a logic level converter.

Now that I know what all the electronic components will be, I can carry on with the layout design for the parts that will be inside the pressure box, how the outside bits will link to them, and get back to building the organ.

Thank you so, so much for sharing all your clever work and for your patient support.

**Reply by bixb922** [reply]
Marvelous! Thank you for the patience. Looks very good!  Please post your progress, either here or in the Busker forum, I look forward to that!

Happy cranking!

**Reply by viragored** [reply]
At some point I'll be starting a new thread in the Busker forum for the MIDI side of my project. My current version of the pressure box base is the fourth - if that turns out to be the final layout that will be a good moment to start the new thread.

I'll continue with my existing thread for the aspects that follow John Smith's plans.  See you over on the forum...

Happy cranking to you too. Which reminds me, I must find someone to sell me some 8mm bright steel rod for the crank shaft. That's hard to find near where I live.

**Reply by bixb922** [reply]
>8mm bright steel rod for the crank shaft.

I live in Santiago, Chile. No local supplier either. I ordered the steel rod and the springs from McMaster-Carr https://www.mcmaster.com/

They have a very good online catalog.

**Reply by viragored** [reply]
Thanks for the link. I've just found an NZ retailer of "Mechatronics". That term seems to cover anything to do with making 3D printers, robots, and other niche hobbies that use electronics. A lucky find, and now I'm ordering the crank rod from there.
They also supply GT2 belt pulleys. They only have 40 teeth pulleys that will fit the sensor. Do you have any recommendations for the number of teeth I should put on the crankshaft? As their stocks are low, I've ordered one each of 20, 40 and 60 teeth for the crank.

**Reply by bixb922** [reply]
That's great that you found a local retailer.  Makes everything easier.

>GT2 belt pulleys

I selected 20 teeth GT2 pulleys and they work well. The force needed for the sensor is minimal. I think that 40 or 60 teeth will work equally well.

**Reply by viragored** [reply]
Is it best to have the same tooth count on crank and sensor, so that their revolutions match?

**Reply by bixb922** [reply]
There is a small advantage to put, say, 60t on the shaft and 20t on the sensor.

If the sensor counts 100 pulses per revolution, now it will count 300 pulses per rev. That increases the precision of the measurement.

**Reply by viragored** [reply]
Thanks again. The "Mechatronics" retailer has only 40T pulleys for the 6mm sensor shaft. The largest for the 8mm crank is 60T, so I'll use that to start with. If it turns out to be worth doing, I should be able to 3D print some pulleys in different sizes. A while back I used a printed idler pulley which worked satisfactorily.
It seems that, with a bit of work and a 3D printed jig, it's possible to make custom sized continuous loop GT2 belts which may be useful to experiment with different sizes. Here's an example: https://www.printables.com/model/417805-gt2-belt-jig-diy-closed-loop-belts
